### PR TITLE
[WIP] Introduce Error Confirmations

### LINKED ISSUE #1865: [Bug] Replacing a machine in pause state will unpause
#### Describe the issue
Replacing a machine will unpause the game

#### Steps to Reproduce
1. Build a machine room, eg Inflator
2. Pause
3. Replace the machine

#### Expected Behavior
Game stays paused

#### System Information
0ef7cb0
macOS 10.16.7
CD

### PR DESCRIPTION
**Pull request by lewri**
*Fixes #1865*

**Describe what the proposed change does**
- Separate out recoverable errors in the game from regular UIConfirmDialog boxes.
- Introduces a system/forced pause when these appear.

### CODE REVIEW

File: CorsixTH/Lua/world.lua [hunk 1]
```diff
@@ -971,16 +971,30 @@ end
...
+--! Sets whether the game should be forcefully paused
+--!param state (bool)
+function World:systemPause(state)
+  self.system_pause = state
+end
+
+--! Reports the system pause status
+--!return (bool) true is system pause is active, else false
+function World:systemPauseStatus()
+  return self.system_pause
```
Comment (lewri) [inline CorsixTH/Lua/world.lua:991]: Oversight.. should probably be:
```suggestion
  return self.system_pause or false
```
Unless we want an afterload.

File: CorsixTH/Lua/dialogs/error_confirm.lua [hunk 2]
```diff
@@ -0,0 +1,123 @@
+--[[ Copyright (c) 2009 Peter "Corsix" Cawley
```
Comment (Alberth289346) [inline CorsixTH/Lua/dialogs/error_confirm.lua:1]: This line looks wrong unless you live in the past and impersonate someone else :)

File: CorsixTH/Lua/dialogs/error_confirm.lua [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+function UIErrConfirm:cancel()
+  self:close(false)
+end
+
+function UIErrConfirm:ok()
+  self:close(true)
+end
+
+--! Closes the confirm dialog
+--!param ok (boolean or nil) whether to call the confirm callback (true) or cancel callback (false/nil)
```
Comment (Alberth289346) [inline CorsixTH/Lua/dialogs/error_confirm.lua:97]: Name the parameter `confirmed` rather than `ok`?

File: CorsixTH/Lua/dialogs/error_confirm.lua [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE. --]]
+
+--! Clone of UIConfirmDialog. Handles errors requiring a confirmation with a forced pause.
```
Comment (Alberth289346) [inline CorsixTH/Lua/dialogs/error_confirm.lua:21]: 'Clone' is bad by definition. I will have to read the other code as well to understand what you're doing, though, so will come back at this later.

File: CorsixTH/Lua/world.lua [hunk 5]
```diff
@@ -971,16 +971,30 @@ end
...
   if not self:isCurrentSpeed("Pause") then
     self:setSpeed("Pause")
   elseif self.prev_speed then
     self:setSpeed(self.prev_speed)
   end
 end
 
+--! Sets whether the game should be forcefully paused
+--!param state (bool)
+function World:systemPause(state)
```
Comment (Alberth289346) [inline CorsixTH/Lua/world.lua:984]: Name doesn't quite express you're changing it, imho.

File: CorsixTH/Lua/world.lua [hunk 6]
```diff
@@ -971,16 +971,30 @@ end
...
+  self.system_pause = state
+end
+
+--! Reports the system pause status
+--!return (bool) true is system pause is active, else false
+function World:systemPauseStatus()
```
Comment (Alberth289346) [inline CorsixTH/Lua/world.lua:990]: This one isn't clear either, imho.

"system pause status" is a concept, a thing. A function is an action, something it changes, queries or constructs. Give a function a verb to express what it does.

File: CorsixTH/Lua/world.lua (hunk 1 above)
Comment (Alberth289346) [reply CorsixTH/Lua/world.lua:991]: And even with an afterload you should initialize its value for a new game.

Comment (lewri) [reply CorsixTH/Lua/world.lua:991]: Yes my fear is I've been too slapdash on this. Oops.

File: CorsixTH/Lua/dialogs/error_confirm.lua (hunk 4 above)
Comment (Alberth289346) [reply CorsixTH/Lua/dialogs/error_confirm.lua:21]: Too much similar that clone isn't warranted indeed. Either merge both windows, or make a base class for the common parts with 2 derived classes.

**Comment by lewri** [comment]
Closing in favour of other one.

**Pull request closed without merging**